src/Uno.UI/UI/Xaml/Controls/WebView/UnoWKWebView.iOSmacOS.cs

```diff
@@ -649,48 +641,73 @@ public override void DidClose(WKWebView webView)
+					this.Log().Debug($"WKNavigationDelegate.DecidePolicy: NavigationType: {navigationAction.NavigationType} Request:{requestUrl} TargetRequest: {navigationAction.TargetFrame?.Request}");
 				}
 
-				var scheme = navigationAction.Request.Url.Scheme.ToLower();
+				var scheme = requestUrl.Scheme.ToLower();
 
-				if (scheme != "http" && scheme != "https")
+				// Note that the "file" scheme is not officially supported by the UWP WebView (https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.webview.unsupportedurischemeidentified?view=winrt-19041#remarks).
+				// We have to support it here for anchor navigation (as long as https://github.com/unoplatform/uno/issues/2998 is not resolved).
+				var isUnsupportedScheme = scheme != "http" && scheme != "https" && scheme != "file";
+				if (isUnsupportedScheme)
 				{
-					bool cancelled = _onUnsupportedUriSchemeIdentified?.Invoke(navigationAction.Request.Url.ToUri()) ?? false;
+					bool cancelled = _unoWKWebView.OnUnsupportedUriSchemeIdentified(requestUrl);
 
 					decisionHandler(cancelled ? WKNavigationActionPolicy.Cancel : WKNavigationActionPolicy.Allow);
+
+					return;
 				}
-				else
+
+				// The WKWebView doesn't raise navigation event for anchor navigation.
+				// When we detect anchor navigation, we must raise the events (NavigationStarting & NavigationFinished) ourselves.
+				var isAnchorNavigation = GetIsAnchorNavigation();
+				if (isAnchorNavigation)
 				{
-					decisionHandler(WKNavigationActionPolicy.Allow);
+					bool cancelled = _unoWKWebView.OnStarted(requestUrl, stopLoadingOnCanceled: false);
+
+					decisionHandler(cancelled ? WKNavigationActionPolicy.Cancel : WKNavigationActionPolicy.Allow);
+
+					if (!cancelled)
+					{
+						_unoWKWebView.OnNavigationFinished(requestUrl);
+					}
+
+					return;
+				}
+
+				// For all other cases, we allow the navigation. This will results in other WKNavigationDelegate methods being called.
+				decisionHandler(WKNavigationActionPolicy.Allow);
+
+				bool GetIsAnchorNavigation()
+				{
+					// If we navigate to the exact same page but with a different location (using anchors), the native control will not notify us of
+					// any navigation. We need to create this notification to indicate that the navigation worked.
+
+					// To detect an anchor navigation, both the previous and new urls need to match on the left part of the anchor indicator ("#")
+					// AND the new url needs to have content on the right of the anchor indicator.
+					var currentUrlParts = _unoWKWebView._urlLastNavigation?.AbsoluteUrl?.ToString().Split(new string[] { "#" }, StringSplitOptions.None);
+					var newUrlParts = requestUrl?.AbsoluteUri?.ToString().Split(new string[] { "#" }, StringSplitOptions.None);
+
+					return currentUrlParts?.Length > 0
+						&& newUrlParts?.Length > 0
+						&& newUrlParts.Length > 1
```

Comment:
dont need to compare newUrlParts against 0 and then 1

Reply:
This is code that was already there (elsewhere).
I think comparing against 0 is for null checking (because it's `?.Length`)

Reply:
but yeah, we can probably do both at the same time 👍